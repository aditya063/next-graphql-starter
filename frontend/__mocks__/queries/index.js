export {
  MOCK_IS_LOGGED_IN,
  MOCK_CURRENT_USER,
  MOCK_ERROR_CURRENT_USER,
  MOCK_SIGN_UP,
  MOCK_SIGN_IN,
  MOCK_SIGN_OUT,
  MOCK_REQUEST_RESET,
  MOCK_RESET_PASSWORD
} from './user';

export {
  MOCK_CREATE_IDEA,
  MOCK_UPDATE_IDEA,
  MOCK_DELETE_IDEA
} from './idea';
