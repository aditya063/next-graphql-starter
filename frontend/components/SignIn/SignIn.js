import Router from 'next/router';
import { Mutation } from 'react-apollo';
import DisplayError from '../DisplayError/DisplayError';
import * as query from './SignIn.query';
import * as sc from './SignIn.style';

class SignIn extends React.PureComponent {
  state = {
    email: '',
    password: ''
  };

  handleChangeInput = e => this.setState({ [e.target.name]: e.target.value });

  handleSubmitForm = async (e, signIn, client) => {
    e.preventDefault();
    this.setState({ password: '' });
    await signIn();
    await client.resetStore();
  };

  handleError = error => error;

  handleCompleted = () => Router.push({ pathname: '/' });

  handleForgotPassClick = () => Router.push({ pathname: '/requestreset' });

  render() {
    const { email, password } = this.state;
    const isInvalid = email === '' || password === '';

    return (
      <Mutation
        mutation={query.SIGN_IN_MUTATION}
        variables={this.state}
        onError={this.handleError}
        errorPolicy="all"
        onCompleted={this.handleCompleted}
      >
        {(signIn, { loading, error, client }) => (
          <>
            <sc.form
              snapshot="SignIn"
              onSubmit={e => this.handleSubmitForm(e, signIn, client)}
            >
              <fieldset disabled={loading} aria-busy={loading}>
                <h2>Sign In</h2>

                {error && <DisplayError error={error} />}

                <label htmlFor="email">
                  Email
                  <sc.inputText
                    type="email"
                    name="email"
                    placeholder="email"
                    value={email}
                    onChange={this.handleChangeInput}
                  />
                </label>

                <label htmlFor="password">
                  Password
                  <sc.inputText
                    type="password"
                    name="password"
                    placeholder="password"
                    value={password}
                    onChange={this.handleChangeInput}
                  />
                </label>

                <sc.inputBtn
                  value="Sign In"
                  type="submit"
                  disabled={isInvalid}
                />
              </fieldset>
            </sc.form>

            <sc.aForgotPass
              value="Forgot password?"
              onClick={this.handleForgotPassClick}
            >
              Forgot password?
            </sc.aForgotPass>
          </>
        )}
      </Mutation>
    );
  }
}

export default SignIn;
