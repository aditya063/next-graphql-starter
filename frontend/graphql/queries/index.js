export {
  IS_LOGGED_IN,
  CURRENT_USER_QUERY,
  SIGN_UP_MUTATION,
  SIGN_IN_MUTATION,
  SIGN_OUT_MUTATION,
  REQUEST_RESET_MUTATION,
  RESET_PASSWORD_MUTATION
} from './user';

export {
  CREATE_IDEA_MUTATION,
  UPDATE_IDEA_MUTATION,
  DELETE_IDEA_MUTATION
} from './idea';
