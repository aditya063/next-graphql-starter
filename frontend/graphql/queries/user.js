import gql from 'graphql-tag';

export const IS_LOGGED_IN = gql`
  query IS_LOGGED_IN {
    isLoggedIn @client
  }
`;

export const CURRENT_USER = gql`
  query CURRENT_USER {
    currentUser {
      id
      email
      ideas {
        id
        content
      }
      color @client
    }
  }
`;

export const USERS = gql`
  query USERS {
    users {
      id
      email
    }
  }
`;

export const SIGN_UP = gql`
  mutation SIGN_UP(
    $email: String!
    $password: String!
    $confirmPassword: String!
  ) {
    signUp(
      email: $email
      password: $password
      confirmPassword: $confirmPassword
    ) {
      accessToken
      user {
        id
      }
    }
  }
`;

export const SIGN_IN = gql`
  mutation SIGN_IN($email: String!, $password: String!) {
    signIn(email: $email, password: $password) {
      accessToken
      user {
        id
      }
    }
  }
`;

export const SIGN_OUT = gql`
  mutation SIGN_OUT {
    signOut
  }
`;

export const REQUEST_RESET = gql`
  mutation REQUEST_RESET($email: String!) {
    requestReset(email: $email)
  }
`;

export const RESET_PASSWORD = gql`
  mutation RESET_PASSWORD(
    $resetToken: String!
    $password: String!
    $confirmPassword: String!
  ) {
    resetPassword(
      resetToken: $resetToken
      password: $password
      confirmPassword: $confirmPassword
    )
  }
`;
