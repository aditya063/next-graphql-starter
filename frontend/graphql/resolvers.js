export const resolvers = {
  Query: {
    renderedOn: () => (typeof window === 'undefined' ? 'Server' : 'Client')
  },

  // Mutation: {},

  User: {
    color: () => 'green'
  }
};
