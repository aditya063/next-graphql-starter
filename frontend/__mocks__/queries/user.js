import * as query from '../../graphql/queries/user';

export const MOCK_IS_LOGGED_IN = {
  request: { query: query.IS_LOGGED_IN },
  result: { data: { isLoggedIn: true } }
};

export const MOCK_CURRENT_USER = {
  request: { query: query.CURRENT_USER },
  result: {
    data: {
      currentUser: { id: '1', email: 'mock@email.com', color: 'mock color' }
    }
  }
};

export const MOCK_ERROR_CURRENT_USER = {
  request: { query: query.CURRENT_USER },
  error: new Error('mock error')
};

export const MOCK_SIGN_UP = {
  request: {
    query: query.SIGN_UP,
    variables: {
      email: 'mock@email.com',
      password: 'mockpass',
      confirmPassword: 'confirmed mockpass'
    }
  },
  result: { data: { signUp: [{ id: '1' }] } }
};

export const MOCK_SIGN_IN = {
  request: {
    query: query.SIGN_IN,
    variables: { email: 'mock@email.com', password: 'mockpass' }
  },
  result: { data: { signIn: [{ id: '1' }] } }
};

export const MOCK_SIGN_OUT = {
  request: { query: query.SIGN_OUT },
  result: { data: { signOut: true } }
};

export const MOCK_REQUEST_RESET = {
  request: {
    query: query.REQUEST_RESET,
    variables: { email: 'mock@email.com' }
  },
  result: { data: { requestReset: true } }
};

export const MOCK_RESET_PASSWORD = {
  request: {
    query: query.RESET_PASSWORD,
    variables: {
      resetToken: '4e4a8fb6e44ec32642cfa410243652f85885bc72',
      password: 'mockpass',
      confirmPassword: 'mockpass'
    }
  },
  result: { data: { resetPassword: [{ id: '1', email: 'mock@email.com' }] } }
};
