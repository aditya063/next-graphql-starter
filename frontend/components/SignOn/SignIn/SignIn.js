/* eslint-disable react/forbid-prop-types */
/* eslint-disable no-unused-vars */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable no-console */
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import DisplayError from '../../DisplayError/DisplayError';
import * as query from './SignIn.query';
import * as Styled from './SignIn.style';

class SignIn extends React.PureComponent {
  state = {
    email: '',
    password: ''
  };

  handleChangeInput = e => this.setState({ [e.target.name]: e.target.value });

  handleSubmitForm = async (e, signIn, client) => {
    e.preventDefault();
    this.setState({ password: '' });
    await signIn();
    await client.resetStore();
  };

  handleError = error => error;

  handleCompleted = () => this.props.close();

  render() {
    const { email, password } = this.state;
    const isInvalid = email === '' || password === '';

    return (
      <Mutation
        mutation={query.SIGN_IN_MUTATION}
        variables={this.state}
        onError={this.handleError}
        onCompleted={this.handleCompleted}
      >
        {(signIn, { loading, error, client }) => (
          <>
            <Styled.form
              onSubmit={e => this.handleSubmitForm(e, signIn, client)}
            >
              <fieldset disabled={loading} aria-busy={loading}>
                <h2>Sign In</h2>

                {error && <DisplayError error={error} />}

                <label htmlFor="email">
                  Email
                  <input
                    type="email"
                    name="email"
                    placeholder="email"
                    value={email}
                    onChange={this.handleChangeInput}
                  />
                </label>

                <label htmlFor="password">
                  Password
                  <input
                    type="password"
                    name="password"
                    placeholder="password"
                    value={password}
                    onChange={this.handleChangeInput}
                  />
                </label>

                <Styled.submitInputBtn
                  value="Sign In"
                  type="submit"
                  disabled={isInvalid}
                />
              </fieldset>
            </Styled.form>

            <a
              value="Forgot password?"
              css="cursor: pointer;"
              onClick={this.props.requestReset}
            >
              Forgot password?
            </a>
          </>
        )}
      </Mutation>
    );
  }
}

SignIn.propTypes = {
  close: PropTypes.func.isRequired,
  requestReset: PropTypes.func.isRequired
};

export default SignIn;
