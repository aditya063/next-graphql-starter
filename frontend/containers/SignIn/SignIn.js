import SignInForm from '../../components/SignInForm/SignInForm';
import ForgotPassDialog from '../../components/ForgotPassDialog/ForgotPassDialog';
import * as sc from './SignIn.style';

const SignIn = React.memo(() => (
  <sc.SignIn>
    <SignInForm />

    <ForgotPassDialog />
  </sc.SignIn>
));

export default SignIn;
