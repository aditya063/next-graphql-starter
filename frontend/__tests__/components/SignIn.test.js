import { shallow, mount } from 'enzyme';
import { ApolloConsumer } from '@apollo/react-components';
import { MockedProvider } from '@apollo/react-testing';
import { InMemoryCache } from 'apollo-cache-inmemory';
import gql from 'graphql-tag';

import SignIn from '../../components/SignIn/SignIn';
import * as query from '../../__mocks__/queries/user';
import resolvers from '../../graphql/resolvers';
import { load } from '../../utils/testing';

describe('SignIn', () => {
  let mockProps;
  let mockQueries;
  let wrapper;

  beforeEach(() => {
    mockProps = {};
    mockQueries = [query.MOCK_SIGN_IN_MUTATION];
    wrapper = shallow(<SignIn {...mockProps} />, {
      disableLifecycleMethods: true,
      wrappingComponent: MockedProvider,
      wrappingComponentProps: { mocks: mockQueries, addTypename: false }
    });
  });

  afterEach(() => jest.resetAllMocks());

  it('renders correctly', () => {
    const wrapperDive = wrapper
      .dive()
      .dive()
      .dive();

    expect(wrapperDive).toMatchSnapshot();
  });

  xit('fires signIn and updates cache after done', async () => {
    let apolloClient;

    const cache = new InMemoryCache();

    wrapper = mount(
      <ApolloConsumer>
        {client => {
          apolloClient = client;
          return <SignIn {...mockProps} />;
        }}
      </ApolloConsumer>,
      {
        disableLifecycleMethods: true,
        wrappingComponent: MockedProvider,
        wrappingComponentProps: {
          addTypename: false,
          mocks: mockQueries,
          cache,
          resolvers
        }
      }
    );

    load(wrapper);

    const email = wrapper.find('input[name="email"]');
    email.simulate('change', {
      target: { name: 'email', value: 'mock@email.com' }
    });

    const password = wrapper.find('input[name="password"]');
    password.simulate('change', {
      target: { name: 'password', value: 'mockpass' }
    });

    load(wrapper);

    console.log(wrapper.debug());

    console.log(wrapper.find('SignIn').state());

    cache.writeData({ data: { isLoggedIn: false } });

    const form = wrapper.find('form');
    console.log(': form', form.debug());

    await form.prop('onSubmit')({ preventDefault: jest.fn() });

    // await load(wrapper, 100);

    // console.log('cache: ', cache.extract());
    console.log('cache: ', apolloClient.cache.extract());

    // check to make sure the cache's contents have been updated
    const response = await cache.readQuery({
      query: gql`
        query IS_LOGGED_IN {
          isLoggedIn @client
        }
      `
    });

    // const response = await cache.readQuery({ query: query.MOCK_IS_LOGGED_IN });

    console.log(': response', response);

    // expect(isLoggedIn).toBeTruthy();
  });
});
