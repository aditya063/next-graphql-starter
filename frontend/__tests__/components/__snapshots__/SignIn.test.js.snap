// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn renders correctly 1`] = `
<Fragment>
  <formstyle__form
    onSubmit={[Function]}
  >
    <fieldset
      aria-busy={false}
      disabled={false}
    >
      <h2>
        Sign In
      </h2>
      <label
        htmlFor="email"
      >
        Email
        <SignInstyle__inputText
          name="email"
          onChange={[Function]}
          placeholder="email"
          type="email"
          value=""
        />
      </label>
      <label
        htmlFor="password"
      >
        Password
        <SignInstyle__inputText
          name="password"
          onChange={[Function]}
          placeholder="password"
          type="password"
          value=""
        />
      </label>
      <SignInstyle__inputBtn
        disabled={true}
        type="submit"
        value="Sign In"
      />
    </fieldset>
  </formstyle__form>
  <SignInstyle__aForgotPass
    value="Forgot password?"
  >
    Forgot password?
  </SignInstyle__aForgotPass>
</Fragment>
`;
